### DISCUSSION TITLE
Concatenate mogrify montage convert composite

### DISCUSSION THREAD

**Discussion by Build-0-Matic** [ImageMagick/ImageMagick#5367] [Help] [answered]
I'm currently running ImageMagick 7.1.0-43-Q16-HDRI-x64 on Windows 10

I am trying to update an old batch script that was originally created with a legacy version of Imagemagick for windows (when the commands were separate).
But I'm having a bit of trouble with it because I now need to add a function.
The script creates an image index page with the name of the folder as the title. In the original script I mogrify all the images into some .mpc thumbs, montage them in a grid and then convert the whole thing into a single JPG that gets added to the original folder.
But now, I need to add a PNG logo in the northwest corner, and that is giving me a bit of a headache. Also, I'd really like to accelerate the script by not relying on a temp folder if it is at all possible. But I really have no idea where to start with this. I tried a few things but only got blob errors... not fun.
Here's the script (I removed a lot of visual junk that was needed to guide the user in the original).
`@echo off
chcp 65001
rem Verify what has been dropped file or folder or nothing
for /f "tokens=1,2 delims=d" %%A in ("-%~a1") do if "%%B" neq "" (
  set finalfolder=%~1
  set type=folder
) else if "%%A" neq "-" (
  set finalfolder=%~dp1
  set type=file
) else (
  set type=nex
)

set PaperFormat=Letter
set paperh=3300
set paperw=2550
set BannerStyle="BannerLarge.png"
set Tile=6x5
set Thumbnail=296x186
set Extent=2550x3300

rem Goto image folder
cd /d "%~1"
rem Get current dir name
for %%* in (.) do set CurrDirName=%%~n*
rem Make temporary folder
md "%temp%\Contact"
echo.
rem Do image manipulations
rem convert all images to JPEG thumbnails
magick mogrify -format mpc -thumbnail %Thumbnail% -path "%temp%\Contact" "*.jpg"
rem Make a grid of all the images
magick montage -units PixelsPerInch -density 300 ^
-tile %Tile% -label %%t -font Arial -pointsize 6 -background "white" ^
-fill "black" -geometry %Thumbnail%+2+2 ^
-auto-orient "%temp%\Contact\*.mpc" -title "%CurrDirName%" "%temp%\Contact\%##Contact.mpc"
rem ADD EXTRA TEXT AND BANNER FUNCTIONS HERE
magick convert "%temp%\Contact\##Contact.mpc" -gravity north -background white -extent %Extent% "%~1\Contact Sheet.jpg"
echo.
rem cleanup workspace
del /Q "%temp%\Contact\*.mpc"
del /Q "%temp%\Contact\*.cache"
rd /S /Q "%temp%\Contact"
pause
`

Thanks

**Comment by snibgo** [comment, marked as answer]
> I need to add a PNG logo in the northwest corner, ...

In your line ...
```
magick convert "%temp%\Contact##Contact.mpc" -gravity north -background white -extent %Extent% "%~1\Contact Sheet.jpg"
```
... you can add your PNG logo, like this:
```
magick convert "%temp%\Contact##Contact.mpc" -gravity north -background white -extent %Extent% yourlogo.png -gravity NorthWest -compose Over -composite "%~1\Contact Sheet.jpg"
```

> Also, I'd really like to accelerate the script by not relying on a temp folder if it is at all possible. 

Not relying on a temp folder will make no significant difference to performance. The only cost is in making the folder, and deleting it. These are both fast operations.

If you *know* that your current directory has no *.mpc or *.cache files then you could put those files there, and delete them afterwards. But it will make almost no performance difference.

**Reply by Build-0-Matic** [reply]
I didn't think of using the convert line to add the logo. Guess I was complicating things for no reason.
I'm really surprised that there is such a negligible difference between doing things in ram or on the drive. But since they are relatively small files and simple operations, using the temp folder does make things extra simple in a sense. 
Thanks!

**Comment by fmw42** [comment]
On IM 7, use magick, not magick convert.

**Comment by snibgo** [comment]
> I'm really surprised that there is such a negligible difference between doing things in ram or on the drive.

I was considering only the difference between the current directory and "%temp%\Contact". If those are on different type of device (eg one is SSD and the other is HD) then there will be a difference.

The "magick montage" and "magick convert" command yu showed could be combined for performance. But not if yo change you "magick convert" as I suggested.

You might get a performance increase if you pipe the output of "magick montage" to the input of "magick convert", so there is no need for "%temp%\Contact##Contact.mpc".

Incidentally, there seem to be bugs in your script, with missing backslashes between `Contact` and file names.

**Reply by Build-0-Matic** [reply]
I know there are bugs... I've changed things around so much lately that I'll probably have to revert to the original and restart from scratch.
I must admit that at this point I really miss a good IDE that watches out for you. Not being a programmer by trade, I don't have all this stuff in my mind ready to use.
